ocaml/database/db_xml.ml

```diff
@@ -146,15 +146,24 @@ module From = struct
               f acc
           | (_, "table"), [((_, "name"), tblname)] ->
               f (tableset, Table.empty, tblname, manifest)
-          | (_, "row"), ((_, "ref"), rf) :: rest ->
-              (* Remove any other duplicate "ref"s which might have sneaked in there *)
-              let rest = List.filter (fun ((_, k), _) -> k <> "ref") rest in
+          | (_, "row"), rest ->
```

Comment:
Altogether, this code is traversing lists multiple times and is inefficient if these lists are long. But that problem is already pre-existing.

The problem at hand is to find the first "ref" entry and to ignore the others. The solution is to split the list based on the key and we expect to find at least one entry. I think the code achieving this could be kept closer together - a sketch is below:

```
let ref, rest =
  match List.partition (fun ((_, k) -> k = "ref")) rest with
  | r::_, rest -> r,rest
  | [] -> raise ...
```